_bleio/characteristic_buffer.py

```diff
@@ -0,0 +1,140 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+`_bleio.characteristic_buffer`
+=======================================================================
+
+_bleio implementation for Adafruit_Blinka_bleio
+
+* Author(s): Dan Halbert for Adafruit Industries
+"""
+from __future__ import annotations
+from typing import Union
+
+import asyncio
+import queue
+import time
+
+import _bleio.adapter_ as adap
+from _bleio.characteristic import Characteristic
+
+Buf = Union[bytes, bytearray, memoryview]
+
+
+class CharacteristicBuffer:
+    """Accumulates a Characteristic's incoming values in a FIFO buffer."""
+
+    def __init__(
+        self,
+        characteristic: Characteristic,
+        *,
+        timeout: float = 1,
+        buffer_size: int = 64
+    ):
+
+        """Monitor the given Characteristic. Each time a new value is written to the Characteristic
+        add the newly-written bytes to a FIFO buffer.
+
+        :param Characteristic characteristic: The Characteristic to monitor.
+          It may be a local Characteristic provided by a Peripheral Service,
+          or a remote Characteristic in a remote Service that a Central has connected to.
+        :param int timeout:  the timeout in seconds to wait for the first character
+          and between subsequent characters.
+        :param int buffer_size: Size of ring buffer that stores incoming data coming from client.
+          Must be >= 1."""
+        self._characteristic = characteristic
+        self._timeout = timeout
+        self._queue = queue.Queue(buffer_size)
+        characteristic.add_notify_callback(self._notify_callback)
+
+    def _notify_callback(self, data: Buf) -> None:
+        # Add data bytes to queue, one at a time.
+        if self._queue.full():
+            # Discard oldest data to make room.
+            while self._queue.qsize() > len(data):
+                self._queue.get_nowait()
+
+        for b in data:
+            try:
+                self._queue.put_nowait(b)
+            except queue.Full:
```

Comment:
Can you check full with an if statement or are you worried about a race condition?

Reply:
This is the recommended way, I believe, and is the least likely to cause a race.